### PR TITLE
SNOW-929218 Add type coercion support for CaseWhen and Iff

### PR DESCRIPTION
**Pull request by sfc-gh-stan**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-929218

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.

3. Please describe how your code solves the related issue.

   Please write a short description of how your code change solves the related issue.

### CODE REVIEW

File: src/snowflake/snowpark/mock/_functions.py [hunk 1]
```diff
@@ -336,6 +337,10 @@ def mock_to_date(
 
         [x] For all other values, a conversion error is generated.
     """
+
+    if isinstance(column.sf_type.datatype, DateType):
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/_functions.py:341]: When a column of type x gets casted to type x, it shouldn't raise error. Same for the changes below

File: src/snowflake/snowpark/mock/_functions.py [hunk 2]
```diff
@@ -1130,6 +1147,10 @@ def convert_num_to_bool(x: Optional[Real]):
         new_col = column.apply(lambda x: try_convert(convert_num_to_bool, try_cast, x))
         new_col.sf_type = ColumnType(BooleanType(), column.sf_type.nullable)
         return new_col
+    elif isinstance(column.sf_type.datatype, VariantType):
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/_functions.py:1150]: Variant to Boolean is supported.

File: src/snowflake/snowpark/mock/_functions.py [hunk 3]
```diff
@@ -1829,3 +1817,63 @@ def mock_concat_ws(*columns: ColumnEmulator) -> ColumnEmulator:
...
+        res = ColumnEmulator(data=[None] * len(condition), dtype=object)
+        expr1 = cast_column_to(expr1, coerce_result)
+        expr2 = cast_column_to(expr2, coerce_result)
+        res.where(condition, other=expr2, inplace=True)
+        res.where([not x for x in condition], other=expr1, inplace=True)
+        res.sf_type = coerce_result
+        return res
+    else:
+        raise SnowparkLocalTestingException(
+            f"[Local Testing] does not support coercion currently, iff expr1 and expr2 have conflicting data types: {expr1.sf_type} != {expr2.sf_type}"
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/_functions.py:1878]: Update this

File: src/snowflake/snowpark/mock/_functions.py [hunk 4]
```diff
@@ -1296,7 +1282,7 @@ def convert_variant_to_array(val):
         res = expr.apply(lambda x: try_convert(convert_variant_to_array, False, x))
     else:
         res = expr.apply(lambda x: try_convert(lambda y: [y], False, x))
-    res.sf_type = ColumnType(ArrayType(), expr.sf_type.nullable)
+    res.sf_type = ColumnType(ArrayType(expr.sf_type.datatype), expr.sf_type.nullable)
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/_functions.py:1285]: Remove this

File: src/snowflake/snowpark/mock/_functions.py [hunk 5]
```diff
@@ -1339,7 +1325,9 @@ def convert_variant_to_object(val):
             f"Invalid type {type(expr.sf_type.datatype)} parameter 'TO_OBJECT'"
         )
 
-    res.sf_type = ColumnType(MapType(), expr.sf_type.nullable)
+    res.sf_type = ColumnType(
+        MapType(VariantType(), VariantType()), expr.sf_type.nullable
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/_functions.py:1329]: Remove this

File: src/snowflake/snowpark/mock/_plan.py [hunk 6]
```diff
@@ -1795,9 +1791,16 @@ def calculate_expression(
                 new_column[
                     left.apply(lambda x: x is not None and np.isnan(x))
                     & right.apply(lambda x: x is not None and np.isnan(x))
-                ] = True
-                # NaN == NaN evaluates to False in pandas, but True in Snowflake
+                ] = True  # NaN == NaN evaluates to False in pandas, but True in Snowflake
                 new_column[new_column.isna() | new_column.isnull()] = False
+            # TODO: the special case with (1,2,3) == [1,2,3] should be True
```
Comment (sfc-gh-stan) [inline src/snowflake/snowpark/mock/_plan.py:1796]: Remove this

File: src/snowflake/snowpark/mock/_snowflake_data_type.py [hunk 7]
```diff
@@ -234,6 +242,74 @@ def calculate_type(c1: ColumnType, c2: Optional[ColumnType], op: Union[str]):
     )
 
 
+def coerce_t1_into_t2(t1: DataType, t2: DataType) -> Optional[DataType]:
```
Comment (sfc-gh-jrose) [inline src/snowflake/snowpark/mock/_snowflake_data_type.py:245]: This function looks mostly correct to me. I suspect we may find additional rules in the future. Can you leave a comment on how you found and verified that these rules were true?

File: src/snowflake/snowpark/mock/_snowflake_data_type.py [hunk 8]
```diff
@@ -234,6 +242,74 @@ def calculate_type(c1: ColumnType, c2: Optional[ColumnType], op: Union[str]):
...
+            return t2
+    elif isinstance(t1, MapType):
+        if isinstance(t2, MapType):
+            if t1.key_type == t2.key_type and t2.value_type == t1.value_type:
+                return t2
+            else:
+                return MapType(key_type=VariantType(), value_type=VariantType())
+        elif isinstance(t2, VariantType):
+            return t2
+    elif isinstance(t1, (TimeType, TimestampType, MapType, ArrayType)):
```
Comment (sfc-gh-jrose) [inline src/snowflake/snowpark/mock/_snowflake_data_type.py:298]: Logically I think TimeType -> TimestampType and DateType -> TimestampType should be valid as well, but I'm not sure if that holds up in practice.

**Review by sfc-gh-jrose — APPROVED** [review]
Awesome PR 🎉 

Are these additional mock unit tests that should be added to more directly test the new helper functions?

**Review by sfc-gh-aalam — APPROVED** [review]

File: src/snowflake/snowpark/mock/_snowflake_data_type.py (hunk 8 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/mock/_snowflake_data_type.py:302]: `DateType -> TimestampType` is supported on line 284, `TimeType -> TimestampType` is not supported by my tests.

**Pull request merged**